Lib/fontmake/instantiator.py

```diff
@@ -0,0 +1,579 @@
+) -> List[Tuple[Location, FontMathObject]]:
+    """Return master Info objects wrapped by MathInfo."""
+    locations_and_masters = []
+    for source in designspace.sources:
+        if source.layerName is not None:
+            continue  # No font info in source layers.
+
+        normalized_location = varLib.models.normalizeLocation(
+            source.location, axis_bounds
+        )
+        locations_and_masters.append(
+            (normalized_location, fontMath.MathInfo(source.font.info))
+        )
+
+    return locations_and_masters
+
+
+def collect_kerning_masters(
+    designspace: designspaceLib.DesignSpaceDocument, axis_bounds: AxisBounds
+) -> List[Tuple[Location, FontMathObject]]:
+    """Return master kerning objects wrapped by MathKerning."""
+    locations_and_masters = []
+    for source in designspace.sources:
+        if source.layerName is not None:
+            continue  # No kerning in source layers.
+
+        # This assumes that groups of all sources are the same.
+        normalized_location = varLib.models.normalizeLocation(
+            source.location, axis_bounds
+        )
+        locations_and_masters.append(
+            (
+                normalized_location,
+                fontMath.MathKerning(source.font.kerning, source.font.groups),
+            )
+        )
+
+    return locations_and_masters
+
+
+def collect_glyph_masters(
+    designspace: designspaceLib.DesignSpaceDocument,
+    glyph_name: str,
+    axis_bounds: AxisBounds,
+) -> List[Tuple[Location, FontMathObject]]:
+    """Return master glyph objects for glyph_name wrapped by MathGlyph.
+
+    Note: skips empty source glyphs if the default glyph is not empty to almost match
+    what ufoProcessor is doing. In e.g. Mutator Sans, the 'S.closed' glyph is left
+    empty in one source layer. One could treat this as a source error, but ufoProcessor
+    specifically has code to skip that empty glyph and carry on.
+    """
+    locations_and_masters = []
+    default_glyph_empty = False
+    other_glyph_empty = False
+
+    for source in designspace.sources:
+        if source.layerName is None:  # Source font.
+            source_layer = source.font.layers.defaultLayer
+        else:  # Source layer.
+            source_layer = source.font.layers[source.layerName]
+
+        # Sparse fonts do not and layers may not contain every glyph.
+        if glyph_name not in source_layer:
+            continue
+
+        source_glyph = source_layer[glyph_name]
+
+        if not (source_glyph.contours or source_glyph.components):
+            if source is designspace.findDefault():
+                default_glyph_empty = True
+            else:
+                other_glyph_empty = True
+
+        # TODO: What to do about a glyphs lib key? Any useful data that should be
+        # copied from master to all instances? `public.verticalOrigin`, does it
+        # interpolate? `public.postscript.hints`?
```

Comment:
yes, public.verticalOrigin should be interpolated. Maybe open an issue in fontmath about that

Reply:
Done: https://github.com/robotools/fontMath/issues/161

Still drop glyph lib key for now?

Reply:
Yes.